src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/AbstractRazorDelegatingEndpoint.cs

```diff
@@ -0,0 +1,140 @@
+
+        /// <summary>
+        /// Override to provide behavior for razor contexts. Since the c#/html servers won't
+        /// have any information for this context it doesn't need to be delegated
+        /// </summary>
+        protected abstract Task<TResponse?> HandleInRazorAsync(TRequest request, DocumentContext documentContext, RazorCodeDocument codeDocument, SourceText sourceText, CancellationToken cancellationToken);
+
+        /// <summary>
+        /// The delegated object to send to the <see cref="EndpointName"/>
+        /// </summary>
+        protected abstract TDelegatedParams CreateDelegatedParams(TRequest request, DocumentContext documentContext, Projection projection);
+
+        /// <summary>
+        /// The name of the endpoint to delegate to, from <see cref="LanguageServerConstants"/>
+        /// </summary>
+        protected abstract string EndpointName { get; }
+
+        /// <summary>
+        /// If the response needs to be remapped for any reasons, override to handle the remapping logic
+        /// </summary>
+        protected virtual Task<TResponse> RemapResponseAsync(TResponse delegatedResponse, RazorCodeDocument codeDocument, CancellationToken cancellationToken)
+            => Task.FromResult(delegatedResponse);
+
+        /// <summary>
+        /// Extracts information about the request and then calls <see cref="GetDelegatedResponseAsync(TRequest, DocumentContext, RazorCodeDocument, SourceText, int, Projection, CancellationToken)"/>.
+        /// Children can override to provide any more custom logic needed before doing a normal delegated response pass with <see cref="GetDelegatedResponseAsync(TRequest, DocumentContext, RazorCodeDocument, SourceText, int, Projection, CancellationToken)"/>
+        /// </summary>
+        public virtual async Task<TResponse?> Handle(TRequest request, CancellationToken cancellationToken)
+        {
+            if (request is null)
+            {
+                throw new ArgumentNullException(nameof(request));
+            }
+
+            var documentContext = await _documentContextFactory.TryCreateAsync(request.TextDocument.Uri, cancellationToken).ConfigureAwait(false);
+            if (documentContext is null)
+            {
+                return null;
+            }
+
+            var codeDocument = await documentContext.GetCodeDocumentAsync(cancellationToken);
+            if (codeDocument.IsUnsupported())
+            {
+                return null;
+            }
+
+            var sourceText = await documentContext.GetSourceTextAsync(cancellationToken).ConfigureAwait(false);
+            if (!request.Position.TryGetAbsoluteIndex(sourceText, _logger, out var absoluteIndex))
+            {
+                return null;
+            }
+
+            var projection = await _documentMappingService.GetProjectionAsync(documentContext, absoluteIndex, cancellationToken).ConfigureAwait(false);
```

Comment:
Given the projection bit already takes in the document context would it make sense to add an overload / extension method for `GetProjectionAsync` that takes in an LSP position?

Reply:
I'm ashamed I didn't think of that (though it did take me a few seconds to work out what this comment meant :P)

Reply:
I added this, at least my understanding of this? LMK if it is wrong